opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/AbstractDaoHibernate.java

```diff
@@ -259,8 +262,31 @@ public void deleteAll(final Collection<T> entities) throws DataAccessException {
 
     @Override
     public List<T> findMatching(final org.opennms.core.criteria.Criteria criteria) {
-        final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
-        return getHibernateTemplate().execute(callback);
+        if(criteria.isMultipleAnd()){
+            Collection<Restriction> allRestrictions = criteria.getRestrictions();
+            Collection<Restriction> copyOfAllRestriction = new ArrayList<Restriction>(allRestrictions) ;
+            List<T> allRecords = new ArrayList<>();
```

Comment:
This can be simplified by breaking restrictions into two sets, one with multiAnd restrictions other without those multiAnd restrictions before looping over.

Reply:
You can use `Sets.difference` from guava to separate multiAnd and  non-multiAnd restrictions.
Use `Sets.intersection` to get intersection of  results from different queries.

Reply:
implementation updated using Sets.difference and Sets.intersection